pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,228 @@
+		SpanID:  t.spanID,
+	}
+}
+
+const maxTraces = 10
+
+// RaftTracer is a utility to trace raft messages. Note that it may include
+// additional messages since it does not consider the term, however it should
+// not miss useful messages.
+//
+// An example higher latency case when this additional info might be useful:
+// * We submit a proposal which gets appended as (term=10,index=100), but this
+// was racing with a leader change.
+// * The new leader at term=11 has overwritten this entry with
+// (term=11,index=100). We will see this in the trace (there should be a bunch
+// of MsgApp/MsgStorageAppend events overlapping this index again).
+// * This causes the leaseholder to wait, and eventually realize that the
+// proposal never popped out committed. It submits a reproposal, which now ends
+// up at (term=11,index=110). We also should see that in the trace (the
+// reproposal will inherit the same trace ID etc).
+type RaftTracer struct {
+	m             syncutil.Map[kvpb.RaftIndex, traceValue]
+	numRegistered atomic.Int64
+	ctx           context.Context
+}
+
+func NewRaftTracer(ctx context.Context) *RaftTracer {
+	return &RaftTracer{ctx: ctx}
+}
+
+// RegisterRemote is used to register a remote trace. This is used when we
+// receive a raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(
+	ctx context.Context,
+	index kvpb.RaftIndex,
+	traceID tracingpb.TraceID,
+	remoteSpanID tracingpb.SpanID,
+) {
+	if r.numRegistered.Load() > maxTraces {
+		return
+	}
+	r.numRegistered.Add(1)
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	t := newRemoteTrace(r.ctx, traceID, remoteSpanID)
+	t.logf(1, "trace index (i%d) from remote %s", index, tracingID(traceID, remoteSpanID))
+	r.m.Store(index, &t)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// Register is called on an entry that is about to be proposed. This will begin
+// logging all subsequent updates to this entry.
+func (r *RaftTracer) Register(baseCtx withContext, ent raftpb.Entry) {
+	if r.numRegistered.Load() > maxTraces {
+		return
+	}
+	r.numRegistered.Add(1)
+	// We store all the entries in the map. This is because we don't know which
+	// one was traced. Typically this is called with only a single entry.
+	t := newLocalTrace(r.ctx, baseCtx)
+	// Log the identical entry to both to make it easier to match up later.
+	t.logf(1, "registering entry (i%d) for tracing", ent.Index)
+	r.m.Store(kvpb.RaftIndex(ent.Index), &t)
+}
+
+// IndexContext is a pair of a raft index and its tracing context.
+type IndexContext struct {
+	Index uint64
+	Ctx   context.Context
+}
+
+// MaybeTrace will log the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []*kvserverpb.TracedEntry {
```

Comment:
Return `TracedEntry` by value, to avoid extra allocations?

Might as well make sure that in the `RaftMessageRequest` these are stored by value, if not already.

Reply:
TIL how to make it a non-pointer in the protobuf for repeated types. Updated!